Show photos on presentation page

mohammedhemaid opened the pull request:
The video of the feature:
https://photos.app.goo.gl/8Q5bR4xXiUooNEQF9

I created a presentation manager that has the animations with the player 

-For the text size and color dialogs, so should I remove them or we need to use them in some way?

mohammedhemaid committed:
Show photos on presentation page

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationActivity.java:
```diff
@@ -85,19 +79,17 @@ protected void onCreate(Bundle savedInstanceState) {
 
+        presentationManager = new PresentationManager(this, namesList);
+        DividerItemDecoration itemDecorator =
+                new DividerItemDecoration(this, DividerItemDecoration.VERTICAL);
```
I am unsure how to feel about the divider. I know it's in the dialog, but on this page, it just seems... off.

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationActivity.java:
```diff
@@ -106,7 +98,6 @@ protected void onCreate(Bundle savedInstanceState) {
                 .onPositive((@NonNull MaterialDialog dialog, @NonNull DialogAction which) -> {
                     int newTextSize = setTextViewHolder.textSizeSlider.getProgress() + 1;
                     preferencesManager.setPresentationTextSize(newTextSize);
-                    names.setTextSize(TypedValue.COMPLEX_UNIT_SP, newTextSize * 8);
```
Users get mad when you disable random features. I think we should make the text size at the adapter layer dynamic and call notifyDataSetChanged() when this happens.

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationManager.kt:
```diff
@@ -0,0 +1,100 @@
+        defaultAnimationLengthMs = view.context.resources.getInteger(R.integer.default_anim_length).toLong()
+    }
+
+    fun playSoundOrAnimations() {
```
playSoundOrAnimations() reveals too much about the implementation, which isn't necessary. Also, if we change the nature of the presentation, we will have to rename this method. To keep it "magic"-sounding, we should just follow the name of the class. I think presentChosenName() is a fitting method name for a PresentationManager class.

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationManager.kt:
```diff
@@ -0,0 +1,100 @@
+        } catch (ex: Exception) {
+            UIUtils.showLongToast(R.string.drumroll_error, view.context)
+        }
+        playNamesAnimation(2600)
```
nit: Let's move 2600 to some sort of constant like ANIMATION_DELAY_MS. Maybe have a comment saying that 2.6s is the amount of time you need to wait for the drumroll crash to happen (we want to sync the animating in of the names with that crash).

Gear61 commented:
The animation looks kind of... weird as well. It looks like it pops out too much and goes really up. Maybe we should tone down the scaling to make it more subtle? Like a 10% or 25% scale.

mohammedhemaid committed:
Change animation scale for presentation, removed RecycleView divider and enabled change text size and color

mohammedhemaid commented:
The video of the changes:
https://photos.app.goo.gl/EF7hHfSPbqtzJ97P6

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationActivity.java:
```diff
@@ -251,7 +176,7 @@ public void onPause() {
         super.finish();
         overridePendingTransition(0, R.anim.slide_out_from_top);
-        player.stop();
+        presentationManager.stopPlayer();
```
nit: stopPlayer() made more sense when we were inlining the MediaPlayer directly. I think stopPresentation() is a better API name given the name.

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationAdapter.kt:
```diff
@@ -0,0 +1,87 @@
+
+        fun loadName(position: Int) {
+            val nameDO: NameDO = chosenNames[position]
+            if (showAsList) {
```
The vast majority of users won't have showAsList setting (it's a pretty niche setting). This adds more code complexity, but let's change the gravity of the container based on this setting to adhere to the old implementation. So if this setting is on, we keep the regular align start since we need to align the numbers. If it's off, set the gravity of the content inside to be centered horizontally. It looks really weird to only select 1 thing and have it off to the side.

mohammedhemaid replied:
yes, it looks very weird thanks for the feedback

Gear61 reviewed (changes requested):
The v2 looks a lot better. On a side note, all this nuance shows how allowing for user customization is a slippery slope that leads to hard to maintain code... I probably accepted too many of these requests.

mohammedhemaid committed:
Set presentation name gravity to center on not showing as list

mohammedhemaid committed:
Set presentation name next to it image

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationManager.kt:
```diff
@@ -0,0 +1,98 @@
+import com.randomappsinc.studentpicker.R
+import com.randomappsinc.studentpicker.utils.UIUtils
+
+/** The amount of time you need to wait for the drumroll crash to happen*/
```
nit: Is it not possible to have this within the class?
```suggestion
/** The amount of time you need to wait for the drumroll crash to happen */
```

mohammedhemaid replied:
Hmmm, we can put it inside a companion object but this is much less clean because under the hood when the bytecode is generated, a useless object is created.

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationManager.kt:
```diff
@@ -0,0 +1,98 @@
+    private var player: MediaPlayer = MediaPlayer()
+    private var handler: Handler = Handler()
+    private val animateNamesTask = ::animateNames
+    private var defaultAnimationLengthMs: Long = view.context.resources.getInteger(
```
Isn't it clearly an integer? We don't need values greater than 2 billion.
```suggestion
    private var defaultAnimationLengthMs: Integer = view.context.resources.getInteger(
```

mohammedhemaid replied:
The ObjectAnimator duration is taking long, Kotlin is weird it doesn't have an implicit casting like java I think that because it doesn't have a primitive data type.

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationManager.kt:
```diff
@@ -0,0 +1,98 @@
+        fullSet.addListener(object : Animator.AnimatorListener {
+            override fun onAnimationStart(animator: Animator) {}
+            override fun onAnimationEnd(animator: Animator) {
+                listener.speakNames()
```
This doesn't seem right - Automatically speaking the names is a toggle-able setting right? Don't we need to read from some boolean?

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationActivity.java:
```diff
@@ -123,85 +111,22 @@ public void choose() {
-            UIUtils.showLongToast(R.string.drumroll_error, this);
+    @Override
+    public void speakNames() {
+        if (settings.getAutomaticTts()) {
```
Ahhh, the speaking check is here. Calling speakNames() to have to sometimes not speak names is confusing and wastes a method call. I think we should just pass the boolean into presentation manager.

Gear61 reviewed (changes requested):
Just some small things - This looks great overall!

mohammedhemaid committed:
Refactor name presentation dialog

mohammedhemaid committed:
Fix color chooser dialog and change say names icon

mohammedhemaid commented:
The video of the changes:
https://photos.app.goo.gl/csuMJrxGXMXMp9Ew6

I pushed the first commit without testing then realized that, so suddenly the app crashed while I was testing it so I fixed it.
Also changed the presentation mic icon to speaker icon, IMHO that looks good and better for the user experience

Gear61 commented on app/src/main/java/com/randomappsinc/studentpicker/presentation/PresentationAdapter.kt:
```diff
@@ -0,0 +1,92 @@
+
+        fun loadName(position: Int) {
+            val nameDO: NameDO = chosenNames[position]
+            if (showAsList) {
```
Show as list can't change while the user is on the screen, so technically you can move this into onCreateViewHolder where the view is inflated so you don't need to do it again on this view being recycled and rebound.

Gear61 reviewed (approved):
This has been in review for a long time so merging this in. Great job!
